### ISSUE TITLE
Explore `target` parameter in HDF5DataSet

### ISSUE THREAD

**Issue by gcroci2** [DeepRank/deeprank-core#152]
In `HDF5DataSet` class, `target` parameter usage can be improved. Docs say that it can be `irmsd, lrmsd, fnat, bin, capri_class or DockQ. Defaults to None.`

- [ ] What happens if `target` is `None`? If it's not possible that there is no target (because there is no logic for it in `NeuralNet` class), we should throw an error in this case.
- [ ] What happens if the user sets a target not in the above list? If this is possible, it should be explicitly written in the docs. 
- [x] Now `task` parameter is set in `NeuralNet` class, according to `target` parameter, which is set in `HDF5DataSet` class. I think it's better to move `target` parameter to `NeuralNet` instead of `HDF5DataSet` class. Note also that conceptually, the target should be passed to either a class or function in a later stage, during the training process. This is another reason why I think it would be handy to move the `target` parameter to the `NeuralNet` class.
    - After discussion below: instead move `task` parameter to `HDF5DataSet` class (done in #200)
- [ ] After having decided about the above, and implemented that, remember to add an example code in the readme of `target` parameter usage.

**Comment by cbaakman** [comment]
If target is None, then no ground truth is defined. It means that a model cannot be trained and only testing/predicting should work.
https://github.com/DeepRank/deeprank-core/blob/cb22435bad8c36024aef47cdf377b30fdb5a03ac/deeprankcore/DataSet.py#L334

If the user sets a target not in the above list, then theoretically it should still work, if that target is in the preprocessed data file. The user will have to manually specify a task as 'class' or 'reg' though.

The Dataset class needs the target name to find the ground truth values in the preprocessed data. That's most likely the reason why this variable must be provided to the class. Without it, the load_one_graph method wouldn't be able to produce any ground truth data.

**Comment by DaniBodor** [comment]
Is this parameter all possible options of what is called `"score"` in storage.py?
If so, should it then also be renamed to `labels`, following the same logic as in issue #150?

**Comment by DaniBodor** [comment]
> Now `task` parameter is set in `NeuralNet` class, according to `target` parameter, which is set in `HDF5DataSet` class. I think it's better to move `target` parameter to `NeuralNet` instead of `HDF5DataSet` class. Note also that conceptually, the target should be passed to either a class or function in a later stage, during the training process. This is another reason why I think it would be handy to move the `target` parameter to the `NeuralNet` class.

The `load_one_graph` method of `HDF5DataSet` uses `target` parameter. Because `NeuralNet` uses an `HDF5DataSet` and not vice versa, I am not sure if it is possible to only set the target there.
It is used to create a torch_geometric `Data` object, which is then returned by the `load_one_graph` method. 
I am not yet familiar enough with the code to understand whether this can be moved in its entirety, with depending methods (PreCluster and get) to NeuralNet and whether that even makes sense to do.

Maybe easiest solution to leave it in HDF5DataSet after all?

**Comment by DaniBodor** [comment]
Do I understand correctly that "score" (soon to be "label", see #150) contains all potential `target` values that can be used to train the network on?

**Comment by gcroci2** [comment]
> > Now `task` parameter is set in `NeuralNet` class, according to `target` parameter, which is set in `HDF5DataSet` class. I think it's better to move `target` parameter to `NeuralNet` instead of `HDF5DataSet` class. Note also that conceptually, the target should be passed to either a class or function in a later stage, during the training process. This is another reason why I think it would be handy to move the `target` parameter to the `NeuralNet` class.
> 
> The `load_one_graph` method of `HDF5DataSet` uses `target` parameter. Because `NeuralNet` uses an `HDF5DataSet` and not vice versa, I am not sure if it is possible to only set the target there. It is used to create a torch_geometric `Data` object, which is then returned by the `load_one_graph` method. I am not yet familiar enough with the code to understand whether this can be moved in its entirety, with depending methods (PreCluster and get) to NeuralNet and whether that even makes sense to do.
> 
> Maybe easiest solution to leave it in `HDF5DataSet` after all? In this case, maybe we should move `task` to `HDF5DataSet` as well to keep them together (which I agree should be part of the same class) @gcroci2

I agree, for now let's move task to HDF5DataSet class.

**Comment by DaniBodor** [comment]
is `target` parameter in `NeuralNet.format_output` the same as the `target` meant in this post or is it an unrelated parameter?
- if same, we should load target = self.target in the method
- if different, we should rename it to avoid confusion

**Comment by DaniBodor** [comment]
> If target is None, then no ground truth is defined. It means that a model cannot be trained and only testing/predicting should work.

I still don't understand this completely, but maybe I am missing something.
As I see it, `target` is only set to NeuralNet in case `pretrained_model is None`. If there is no pretrained model, then there is no possibility of testing. In this case, target should still not be allowed to be None? We can still make `target` optional, because we don't want to force a user to set it for a test-only parameter, but we should probably raise an error if it is set to `None` if there is no pretrained model. 
Is this correct?

**Comment by DaniBodor** [comment]
> The `load_one_graph` method of `HDF5DataSet` uses `target` parameter. Because `NeuralNet` uses an `HDF5DataSet` and not vice versa, I am not sure if it is possible to only set the target there. It is used to create a torch_geometric `Data` object, which is then returned by the `load_one_graph` method. I am not yet familiar enough with the code to understand whether this can be moved in its entirety, with depending methods (PreCluster and get) to NeuralNet and whether that even makes sense to do.
> 
> Maybe easiest solution to leave it in `HDF5DataSet` after all? In this case, maybe we should move `task` to `HDF5DataSet` as well to keep them together (which I agree should be part of the same class) @gcroci2

Only realized now that @cbaakman already said more or less the same thing in his comment above

**Comment by gcroci2** [comment]
> is `target` parameter in `NeuralNet.format_output` the same as the `target` meant in this issue or is it a different parameter?

`target`  parameter in `NeuralNet.format_output` method is an array containing target values, while `target` parameter of HDF5DataSet class (the one meant in this issue) is a string, indicating target variable's name. 

> * if same, should we load `target = self.target` in the method?
> * if different, should we rename it to avoid confusion?

I think is fine as it is, it's not the best but format_output it's used only internally in the training loop, and its usage is clear there.

**Comment by gcroci2** [comment]
Now I have a much clearer picture of what's going on regarding target/task among NeuralNet.py and DataSet.py. Let's have a chat together about this issue, and then we'll finalize it @DaniBodor

**Comment by DaniBodor** [comment]
> The Dataset class needs the target name to find the ground truth values in the preprocessed data. That's most likely the reason why this variable must be provided to the class. Without it, the load_one_graph method wouldn't be able to produce any ground truth data.

@cbaakman , @gcroci2 , does either of you know what the purpose of this method is? It is not being called anywhere in the code, so I guess it is intended as a function that a user can call to inspect a graph.

If this is the case, wouldn't it make more sense for the graph to show all available targets & features and not only the ones selected? In PR #228 (issue #194) I plan to move assigning features and targets to the `Trainer` class instead of the `HDF5DataSet` class. I could then either move the `load_one_graph` method along, using only the selected features/targets, or update the function to include all features/targets in the hdf5 file. My gut goes to the latter solution, but as I'm not sure what this function is exactly intended for, I can't be sure.

**Comment by cbaakman** [comment]
`load_one_graph` is essential. It takes graph data out of the hdf5 file, so it also needs to take the ground truth values associated with that graph. However, it doesn't matter whether it's inside the `Trainer` or `HDF5Dataset` class. It's used when loading a dataset though.
If you load all targets from the hdf5 file, then the Trainer needs to specify which target to use.
